File: tasks/expr-eval/typescript/src/evaluate.ts
import { Token } from "./parse";

/** Left to right, with no precedence between the operators. */
export function evaluate(tokens: Token[]): number {
  let value = (tokens[0] as Token).value as number;
  let index = 1;
  while (index < tokens.length) {
    const operator = (tokens[index] as Token).value as string;
    const right = (tokens[index + 1] as Token).value as number;
    value = operator === "+" ? value + right : value * right;
    index += 2;
  }
  return value;
}

File: tasks/expr-eval/go/src/evaluate.go
package main

// evaluate walks the tokens left to right, with no precedence between the
// operators.
func evaluate(tokens []token) int64 {
	value := tokens[0].Number
	index := 1
	for index < len(tokens) {
		operator := tokens[index].Text
		right := tokens[index+1].Number
		if operator == "+" {
			value = value + right
		} else {
			value = value * right
		}
		index += 2
	}
	return value
}
